lib/fluent/plugin/checkpoint_metadata_cache.rb

```diff
@@ -0,0 +1,111 @@
+#
+# Fluentd Kubernetes Metadata Filter Plugin - Enrich Fluentd events with
+# Kubernetes metadata
+#
+# Copyright 2017 Red Hat, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+require "sqlite3"
+
+module KubernetesMetadata
+  module CheckPointCache
+    def start_checkpoint
+      while @checkpoint_thread_running
```

Comment:
Set and compare and actually use the mutex here in line with our discussion; return if already set